In [3]:
import pandas as pd
import numpy as np

In [ ]:
# read in the data
books = pd.read_csv("/Users/kapilojha/Desktop/ml/projects/book-recommender/data/books.csv",
                encoding="latin-1",
                on_bad_lines="skip",
                sep=";",
                low_memory=False
        )

/var/folders/mw/0qwwq_0146j20zhtp_s1sb4c0000gn/T/ipykernel_15867/3863721045.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv("/Users/kapilojha/Desktop/ml/projects/book-recommender/data/books.csv",


In [7]:
!ls

books_data     research.ipynb


In [20]:
# view the data
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [24]:
print("Large URl: ", books["Image-URL-L"].iloc[0])
print("Medium URl: ", books["Image-URL-M"].iloc[0])
print("Small URl: ", books["Image-URL-S"].iloc[0])

Large URl:  http://images.amazon.com/images/P/0195153448.01.LZZZZZZZ.jpg
Medium URl:  http://images.amazon.com/images/P/0195153448.01.MZZZZZZZ.jpg
Small URl:  http://images.amazon.com/images/P/0195153448.01.THUMBZZZ.jpg


Each book has 3 diffrent Image url :
- depending on our requirements we can choose any of the three
- small, medium and large.

In [25]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271358 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


In [27]:
books.shape

(271360, 8)

We have total of 2,71,360 rows and 8 features.

Since we do not need the small and medium url's we should remove them from
our dataset.

In [28]:
books.drop(["Image-URL-S", "Image-URL-M"],
           axis=1,
           inplace=True)
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [29]:
# rename the column names for easier processing
books.rename(columns={
    "ISBN" : "isbn",
    "Book-Title" : "title",
    "Book-Author" : "author",
    "Year-Of-Publication" : "year",
    "Publisher" : "publisher",
    "Image-URL-L" : "image-url"
}, inplace=True)
books.head()

isbn                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author  year                   publisher  \
0    Mark P. O. Morford  2002     Oxford University Press   
1  Richard Bruce Wright  2001       HarperFlamingo Canada   
2          Carlo D'Este  1991             HarperPerennial   
3      Gina Bari Kolata  1999        Farrar Straus Giroux   
4       E. J. W. Barber  1999  W. W. Norton &amp; Company   

                                           image-url  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [41]:
# import users data
users = pd.read_csv("/Users/kapilojha/Desktop/ml/projects/book-recommender/data/users.csv",
                    sep=";",
                    encoding="latin-1",
                    on_bad_lines="skip")
users["Age"] = users["Age"].bfill()
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa  18.0
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia  17.0
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom  61.0

In [42]:
# rename users columns for easier processing
users.rename(columns={
    "User-ID" : "id",
    "Location" : "location",
    "Age" : "age"
}, inplace=True)

In [43]:
users.head()

id                            location   age
0   1                  nyc, new york, usa  18.0
1   2           stockton, california, usa  18.0
2   3     moscow, yukon territory, russia  17.0
3   4           porto, v.n.gaia, portugal  17.0
4   5  farnborough, hants, united kingdom  61.0

In [44]:
# import the book ratings data
ratings = pd.read_csv("/Users/kapilojha/Desktop/ml/projects/book-recommender/data/ratings.csv",
                      sep=";",
                      encoding="latin-1",
                      on_bad_lines="skip")
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [47]:
# rename ratings columns names for easier processing
ratings.rename(columns={
    "User-ID" : "id",
    "ISBN" : "isbn",
    "Book-Rating" : "rating"
}, inplace=True)
ratings.head()

id        isbn  rating
0  276725  034545104X       0
1  276726  0155061224       5
2  276727  0446520802       0
3  276729  052165615X       3
4  276729  0521795028       6

In [48]:
# view the data shapes together
print(books.shape, users.shape, ratings.shape)

(271360, 6) (278858, 3) (1149780, 3)


In [50]:
# no. of books an individual user read
ratings["id"].value_counts()[:10]

id
11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
212898     4785
278418     4533
76352      3367
110973     3100
235105     3067
Name: count, dtype: int64

As you can see `user 11676` has read `13,602 books`.

In [ ]:
# users who have at least read more than 200 books
x = ratings["id"].value_counts() > 200

105283

In [58]:
x.value_counts()

count
False    104384
True        899
Name: count, dtype: int64

We can see only 899 users have read more than 200 books.

In [63]:
# extracting those users id's 
y = x[x].index
y

Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352, 110973,
       235105,
       ...
       260183,  73681,  44296, 155916,   9856, 274808,  28634,  59727, 268622,
       188951],
      dtype='int64', name='id', length=899)

In [68]:
# select only those rows with people who read more than 200 books in ratings dataset
ratings = ratings[ratings["id"].isin(y)]
ratings.head()

id        isbn  rating
1456  277427  002542730X      10
1457  277427  0026217457       0
1458  277427  003008685X       8
1459  277427  0030615321       0
1460  277427  0060002050       0